File: Sort/quick sort/quickSort.java
import java.util.Arrays;

/**
 * quickSort
 */
public class quickSort {

    private static void sort(int arr[]) {
        qSort(arr, 0, arr.length - 1);
    }

    private static void qSort(int arr[], int start, int end) {
        if (start < end) {
            int left = start, right = end;
            int mid = (start + end) >> 1;
            SortTemp.swap(arr, left, mid);
            int temp = arr[left];
            while (left < right) {
                while (left < right && arr[right] <= temp) {
                    right--;
                }
                arr[left] = arr[right];
                while (left < right && arr[left] >= temp) {
                    left++;
                }
                arr[right] = arr[left];
            }
            arr[left] = temp;
            qSort(arr, start, left - 1);
            qSort(arr, left + 1, end);
        }
    }

    public static void main(String[] args) {
        int arr[] = SortTemp.create();
        long startTime = System.currentTimeMillis();
        sort(arr);
        long endTime = System.currentTimeMillis();
        System.out.println("用时:"+ (endTime-startTime)+ " ms");
        // System.out.println(Arrays.toString(arr));
    }
}

File: Sort/quick sort/quickSort.js
function sort(arr) {
    quickSort(arr, 0, arr.length - 1);
}

function quickSort(arr, begin, end) {
    if (begin < end) {

        let left = begin;
        let right = end;
        let mid = (begin + end) >> 1;
        swap(arr, left, mid)
        let t = arr[left];
        while (left < right) {
            while (left < right && t <= arr[right]) { //从后向前找第一个大于t基数的数
                right--;
            }
            arr[left] = arr[right] //将找到的数放到基数位置
            while (left < right && arr[left] <= t) {//从前向后找第一个小于t基数的数
                left++;
            }
            arr[right] = arr[left];////将找到的数放到之前那个right 位置
        }
        arr[left] = t;
        quickSort(arr, begin, left - 1);
        quickSort(arr, left + 1, end);
    }

}


function swap(arr, first, second) {
    let temp = arr[first]
    arr[first] = arr[second]
    arr[second] = temp
}


function create() {
    let arr = []
    for (let index = 0; index < 30; index++) {
        arr.push(Math.floor(Math.random() * 10000 + 1));
    }
    return arr;
}


let array = create()

sort(array)

console.log(array);
